cmake/ProtoBuf.cmake

```diff
@@ -152,22 +166,44 @@ function(caffe2_protobuf_generate_cpp_py srcs_var hdrs_var python_var)
     else()
       set(DLLEXPORT_STR "")
     endif()
-    add_custom_command(
-      OUTPUT "${CMAKE_CURRENT_BINARY_DIR}/${fil_we}.pb.cc"
-             "${CMAKE_CURRENT_BINARY_DIR}/${fil_we}.pb.h"
-             "${CMAKE_CURRENT_BINARY_DIR}/${fil_we}_pb2.py"
-      WORKING_DIRECTORY ${PROJECT_SOURCE_DIR}
-      COMMAND ${CMAKE_COMMAND} -E make_directory "${CMAKE_CURRENT_BINARY_DIR}"
-      COMMAND ${CAFFE2_PROTOC_EXECUTABLE} -I${PROJECT_SOURCE_DIR} --cpp_out=${DLLEXPORT_STR}${PROJECT_BINARY_DIR} ${abs_fil}
-      COMMAND ${CAFFE2_PROTOC_EXECUTABLE} -I${PROJECT_SOURCE_DIR} --python_out "${PROJECT_BINARY_DIR}" ${abs_fil}
-      DEPENDS ${CAFFE2_PROTOC_EXECUTABLE} ${abs_fil}
-      COMMENT "Running C++/Python protocol buffer compiler on ${fil}" VERBATIM )
+
+    if (${CAFFE2_LINK_LOCAL_PROTOBUF})
+      # We need to rewrite the pb.h files to route GetEmptyStringAlreadyInited
+      # through our wrapper in proto_utils so the memory location test
+      # is correct.
+      add_custom_command(
+        OUTPUT "${CMAKE_CURRENT_BINARY_DIR}/${fil_we}.pb.cc"
+               "${CMAKE_CURRENT_BINARY_DIR}/${fil_we}.pb.h"
+               "${CMAKE_CURRENT_BINARY_DIR}/${fil_we}_pb2.py"
+        WORKING_DIRECTORY ${PROJECT_SOURCE_DIR}
+        COMMAND ${CMAKE_COMMAND} -E make_directory "${CMAKE_CURRENT_BINARY_DIR}"
+        COMMAND ${CAFFE2_PROTOC_EXECUTABLE} -I${PROJECT_SOURCE_DIR} --cpp_out=${DLLEXPORT_STR}${PROJECT_BINARY_DIR} ${abs_fil}
+        COMMAND ${CAFFE2_PROTOC_EXECUTABLE} -I${PROJECT_SOURCE_DIR} --python_out "${PROJECT_BINARY_DIR}" ${abs_fil}
+
+        # If we remove all reference to these pb.h files from external
+        # libraries and binaries this rewrite can be removed.
+        COMMAND sed -i.bk -e "s/::google::protobuf::internal::GetEmptyStringAlreadyInited/GetEmptyStringAlreadyInited/g" "${CMAKE_CURRENT_BINARY_DIR}/${fil_we}.pb.h"
```

Comment:
Just curious why do we have this special treatment of `::google::protobuf::internal::GetEmptyStringAlreadyInited`?

Reply:
Good question. It's because it uses a global object to check empty strings. Calling that across library boundaries will fail and create seg faults, so we need to isolate it to one so file.